File: BuddyGrammarKit/Sources/BuddyGrammarKit/SinglePointerInteractionOwner.swift
/// Grants one opaque pointer or gesture token exclusive ownership of an
/// interaction session until that same token releases it.
///
/// Native keyboard adapters use this small, platform-neutral seam to prevent
/// two simultaneous key gestures from driving the same stateful router.
public struct SinglePointerInteractionOwner<Token: Hashable> {
    public private(set) var activeToken: Token?

    public init() {}

    /// Returns `true` only when `token` acquired an idle session.
    @discardableResult
    public mutating func acquire(_ token: Token) -> Bool {
        guard activeToken == nil else { return false }
        activeToken = token
        return true
    }

    public func owns(_ token: Token) -> Bool {
        activeToken == token
    }

    /// Releases ownership only for the active token. A non-owner is ignored.
    @discardableResult
    public mutating func release(_ token: Token) -> Bool {
        guard activeToken == token else { return false }
        activeToken = nil
        return true
    }

    /// Force-clears ownership for adapter teardown or a layout reset.
    public mutating func reset() {
        activeToken = nil
    }
}

File: android/app/src/main/java/com/francescooddo/buddygrammar/core/SinglePointerInteractionOwner.kt
package com.francescooddo.buddygrammar.core

/**
 * Grants one opaque pointer token exclusive ownership of a stateful input
 * session. Move, release, and cancel events from every other token are ignored.
 */
class SinglePointerInteractionOwner<Token : Any> {
    var activeToken: Token? = null
        private set

    fun acquire(token: Token): Boolean {
        if (activeToken != null) return false
        activeToken = token
        return true
    }

    fun owns(token: Token): Boolean = activeToken == token

    fun release(token: Token): Boolean {
        if (!owns(token)) return false
        activeToken = null
        return true
    }

    fun reset() {
        activeToken = null
    }
}
